dali/imgcodec/decoders/decoder_parallel_impl.h

```diff
@@ -74,13 +97,69 @@ class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
     std::vector<DecodeResult> ret(out.size());
     ROI no_roi;
     for (int i = 0; i < in.size(); i++) {
-      tp_->AddWork([&, i](int) {
-        ret[i] = Decode(out[i], in[i], opts, rois.empty() ? no_roi : rois[i]);
+      tp_->AddWork([&, i](int tid) {
+        ret[i] = DecodeImplTask(tid, out[i], in[i], opts, rois.empty() ? no_roi : rois[i]);
+      }, volume(out[i].shape()));
+    }
+    tp_->RunAll();
+    return ret;
+  }
+
+  std::vector<DecodeResult> ParallelDecodeImpl(cudaStream_t stream,
+                                               span<SampleView<GPUBackend>> out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+    assert(out.size() == in.size());
+    assert(rois.empty() || rois.size() == in.size());
+    std::vector<DecodeResult> ret(out.size());
+    ROI no_roi;
+    for (int i = 0; i < in.size(); i++) {
+      tp_->AddWork([&, i](int tid) {
+        ret[i] = DecodeImplTask(stream, tid, out[i], in[i], opts, rois.empty() ? no_roi : rois[i]);
       }, volume(out[i].shape()));
     }
     tp_->RunAll();
     return ret;
   }
+
+  /**
+   * @brief Single image decode CPU implementation, executed in a thread pool context.
+   * 
+   * @param thread_idx thread index in the thread pool
+   * @param out output sample view
+   * @param in encoded image source
+   * @param opts decoding parameters
+   * @param roi region-of-interest
+   * @return std::vector<DecodeResult> 
+   */
+  virtual DecodeResult DecodeImplTask(int thread_idx,
+                                      SampleView<CPUBackend> out,
+                                      ImageSource *in,
+                                      DecodeParams opts,
+                                      const ROI &roi) {
+    throw std::logic_error("Backend not supported");
+  }
+
+  /**
+   * @brief Single image decode GPU implementation, executed in a thread pool context.
+   * 
+   * @param stream CUDA stream to synchronize with
+   * @param thread_idx thread index in the thread pool
+   * @param out output sample view
+   * @param in encoded image source
+   * @param opts decoding parameters
+   * @param roi region-of-interest
+   * @return std::vector<DecodeResult> 
+   */
+  virtual DecodeResult DecodeImplTask(cudaStream_t stream,
+                                      int thread_idx,
```

Comment:
Perhaps here thread_idx should go first - the thread_idx is something added on top of the existing parameters.

Reply:
Agree. I'll fix